src/commands/deploy.js

```diff
@@ -16,20 +17,265 @@ const SitesCreateCommand = require('./sites/create')
+  const funcConfig = get(config, 'build.functions') || get(config, 'build.Functions')
+  if (flags['functions']) {
+    functionsFolder = path.resolve(process.cwd(), flags['functions'])
+  } else if (funcConfig) {
+    functionsFolder = path.resolve(site.root, funcConfig)
+  } else if (get(siteData, 'build_settings.functions_dir')) {
+    functionsFolder = path.resolve(site.root, get(siteData, 'build_settings.functions_dir'))
+  }
+  return functionsFolder
+}
+
+const validateFunctionsFolder = async ({ functionsFolder, log, error }) => {
+  let stat
+  if (functionsFolder) {
+    // we used to hard error if functions folder is specified but doesn't exist
+    // but this was too strict for onboarding. we can just log a warning.
+    try {
+      stat = await statAsync(functionsFolder)
+    } catch (e) {
+      if (e.code === 'ENOENT') {
+        log(
+          `Functions folder "${functionsFolder}" specified but it doesn't exist! Will proceed without deploying functions`
+        )
+      }
+      // Improve the message of permission errors
+      if (e.code === 'EACCES') {
+        error('Permission error when trying to access functions folder')
+      }
+    }
+  }
+
+  if (stat && !stat.isDirectory()) {
+    error('Functions folder must be a path to a directory')
+  }
+
+  return stat
+}
+
+const validateFolders = async ({ deployFolder, functionsFolder, error, log }) => {
+  const deployFolderStat = await validateDeployFolder({ deployFolder, error })
+  const functionsFolderStat = await validateFunctionsFolder({ functionsFolder, error, log })
+  return { deployFolderStat, functionsFolderStat }
+}
+
+const runDeploy = async ({
+  flags,
+  deployToProduction,
+  siteData,
+  api,
+  siteId,
+  deployFolder,
+  configPath,
+  functionsFolder,
+  alias,
+  log,
+  warn,
+  error,
+  exit,
+}) => {
+  let results
+  try {
+    if (deployToProduction) {
+      if (isObject(siteData.published_deploy) && siteData.published_deploy.locked) {
+        log(`\n${NETLIFYDEVERR} Deployments are "locked" for production context of this site\n`)
+        const { unlockChoice } = await inquirer.prompt([
+          {
+            type: 'confirm',
+            name: 'unlockChoice',
+            message: 'Would you like to "unlock" deployments for production context to proceed?',
+            default: false,
+          },
+        ])
+        if (!unlockChoice) exit(0)
+        await api.unlockDeploy({ deploy_id: siteData.published_deploy.id })
+        log(`\n${NETLIFYDEVLOG} "Auto publishing" has been enabled for production context\n`)
+      }
+      log('Deploying to main site URL...')
+    } else {
+      log('Deploying to draft URL...')
+    }
+
+    results = await api.deploy(siteId, deployFolder, {
+      configPath,
+      fnDir: functionsFolder,
+      statusCb: flags.json || flags.silent ? () => {} : deployProgressCb(),
+      draft: !deployToProduction && !alias,
+      message: flags.message,
+      deployTimeout: flags.timeout * 1000 || 1.2e6,
```

Comment:
Should we extract that `1.2e6` into a top-level constant variable?

Reply:
Sure going to update it

Reply:
Done in https://github.com/netlify/cli/pull/1240/commits/f00db3f9b558720907882f3fe2df9e41e276ea9d#diff-e9001bf37e31e33db8a717da3ac59a69R179